python/tritonserver/_c/tritonserver_pybind.cc

```diff
@@ -896,106 +871,128 @@ class PyInferenceRequest
+            request_ptr.release()) /* request_release_userp */));
   }
-
-  static void PyTritonRequestReleaseCallback(
+  static void ReleaseCallback(
       struct TRITONSERVER_InferenceRequest* request, const uint32_t flags,
       void* userp)
   {
-    py::gil_scoped_acquire gil;
-    auto cr = reinterpret_cast<CallbackResource*>(userp);
-    cr->release_fn(cr->request, flags, cr->user_object);
-    delete cr;
+    std::unique_ptr<std::shared_ptr<PyInferenceRequest>> request_ptr(
+        reinterpret_cast<std::shared_ptr<PyInferenceRequest>*>(userp));
   }
 
-  void SetResponseCallback(
-      const py::object& allocator, const py::object& allocater_user_object,
-      PyInferenceResponse::CompleteFn response,
-      const py::object& response_user_object)
+  void SetResponseAllocator()
+  {
+    struct TRITONSERVER_ResponseAllocator* allocator_raw = nullptr;
+    ThrowIfError(TRITONSERVER_ResponseAllocatorNew(
+        &allocator_raw, ResponseAllocatorAllocFn, ResponseAllocatorReleaseFn,
+        nullptr /* start_fn */));
+    allocator_.reset(allocator_raw, TRITONSERVER_ResponseAllocatorDelete);
+    ThrowIfError(TRITONSERVER_ResponseAllocatorSetQueryFunction(
+        allocator_.get(), ResponseAllocatorQueryFn));
+    ThrowIfError(TRITONSERVER_ResponseAllocatorSetBufferAttributesFunction(
+        allocator_.get(), ResponseAllocatorBufferAttributesFn));
+  }
+  static TRITONSERVER_Error* ResponseAllocatorQueryFn(
+      struct TRITONSERVER_ResponseAllocator* allocator, void* userp,
+      const char* tensor_name, size_t* byte_size,
+      TRITONSERVER_MemoryType* memory_type, int64_t* memory_type_id)
+  {
+    *memory_type = TRITONSERVER_MEMORY_CPU;
+    *memory_type_id = 0;
+    return nullptr;
+  }
+  static TRITONSERVER_Error* ResponseAllocatorBufferAttributesFn(
+      struct TRITONSERVER_ResponseAllocator* allocator, const char* tensor_name,
+      struct TRITONSERVER_BufferAttributes* buffer_attributes, void* userp,
+      void* buffer_userp)
   {
-    allocator_callback_resource_.reset(
-        new PyResponseAllocator::CallbackResource(
-            allocator, allocater_user_object));
-    response_callback_resource_.reset(new PyInferenceResponse::CallbackResource(
-        response, allocator_callback_resource_.get(), response_user_object));
+    ThrowIfError(TRITONSERVER_BufferAttributesSetMemoryType(
+        buffer_attributes, TRITONSERVER_MEMORY_CPU));
+    ThrowIfError(TRITONSERVER_BufferAttributesSetMemoryTypeId(
+        buffer_attributes, 0 /* memory_type_id */));
+    return nullptr;
+  }
+  static TRITONSERVER_Error* ResponseAllocatorAllocFn(
+      struct TRITONSERVER_ResponseAllocator* allocator, const char* tensor_name,
+      size_t byte_size, TRITONSERVER_MemoryType memory_type,
+      int64_t memory_type_id, void* userp, void** buffer, void** buffer_userp,
+      TRITONSERVER_MemoryType* actual_memory_type,
+      int64_t* actual_memory_type_id)
+  {
+    *buffer = malloc(byte_size * sizeof(uint8_t));
```

Comment:
are there tritonserver apis we should be using for memory allocation instead of directly calling malloc?

Reply:
I don't think there is one, because there are multiple calls to malloc in the core code base for different purpose, i.e.
https://github.com/triton-inference-server/core/blob/r24.12/src/backend_memory_manager.cc#L98
https://github.com/triton-inference-server/core/blob/r24.12/src/backend_model_instance.cc#L73
https://github.com/triton-inference-server/core/blob/r24.12/src/cache_manager.cc#L95

There is one on the server repo, but it is our of reach from the core binding:
https://github.com/triton-inference-server/server/blob/r24.12/src/memory_alloc.cc#L96

Reply:
@GuanLuo to double check - should we be using the backend_memory_manager ? Reason I ask is that these / and the server already do support the different types of memory gpu / cpu / cpu pinned - and we would want that kind of support here if we can get it - and we should reuse it ...

Reply:
What is in server certainly won't work, it is tightly related to the server frontend for optimization which is why allocator is up for the user to define.

You probably can reuse the one for backend but that is kind of strange. Plus the one in backend may not be as well crafted as you expected.. Please evaluate whether it may sense to proceed in this way.

Reply:
Let's make that a todo as we bring in support for more memory types - we should look to see if we can reuse a common memory allocator in the core - of the ones backend_memeory_manager seems the closest - but agree is odd to reuse here directly.